utils/ixdtf/src/parsers/annotations.rs

```diff
@@ -18,66 +18,67 @@ use crate::{
     ParserError, ParserResult,
 };
 
-use alloc::vec::Vec;
-
 /// Strictly a parsing intermediary for the checking the common annotation backing.
 pub(crate) struct AnnotationSet<'a> {
     pub(crate) tz: Option<TimeZoneAnnotation<'a>>,
-    pub(crate) calendar: Option<&'a str>,
-    pub(crate) annotations: Vec<Annotation<'a>>,
+    pub(crate) calendar: Option<Annotation<'a>>,
 }
 
 /// Parse a `TimeZoneAnnotation` `Annotations` set
-pub(crate) fn parse_annotation_set<'a>(cursor: &mut Cursor<'a>) -> ParserResult<AnnotationSet<'a>> {
+pub(crate) fn parse_annotation_set<'a>(
+    cursor: &mut Cursor<'a>,
+    handler: impl FnMut(Annotation<'a>) -> ParserResult<()>,
+) -> ParserResult<AnnotationSet<'a>> {
     // Parse the first annotation.
     let tz_annotation = timezone::parse_ambiguous_tz_annotation(cursor)?;
 
     // Parse any `Annotations`
     let annotations = cursor.check_or(false, is_annotation_open);
 
     if annotations {
-        let annotations = parse_annotations(cursor)?;
+        let calendar = parse_annotations(cursor, handler)?;
         return Ok(AnnotationSet {
             tz: tz_annotation,
-            calendar: annotations.0,
-            annotations: annotations.1,
+            calendar,
         });
     }
 
     Ok(AnnotationSet {
         tz: tz_annotation,
         calendar: None,
-        annotations: Vec::default(),
     })
 }
 
 /// Parse any number of `KeyValueAnnotation`s
 pub(crate) fn parse_annotations<'a>(
     cursor: &mut Cursor<'a>,
-) -> ParserResult<(Option<&'a str>, Vec<Annotation<'a>>)> {
-    let mut annotations = Vec::default();
-    let mut calendar = None;
-    let mut calendar_crit = false;
+    mut handler: impl FnMut(Annotation<'a>) -> ParserResult<()>,
+) -> ParserResult<Option<Annotation<'a>>> {
+    let mut calendar: Option<Annotation<'a>> = None;
 
     while cursor.check_or(false, is_annotation_open) {
-        let kv = parse_kv_annotation(cursor)?;
+        let kv: Annotation<'a> = parse_kv_annotation(cursor)?;
 
+        // Check if the key is the registered key "u-ca".
         if kv.key == "u-ca" {
-            if calendar.is_none() {
-                calendar = Some(kv.value);
-                calendar_crit = kv.critical;
-                continue;
-            }
-
-            if calendar_crit || kv.critical {
-                return Err(ParserError::CriticalDuplicateCalendar);
+            match calendar {
+                // If any calendar is critical with a value that is not aligned with the set calendar, treat as erroneous and throw an error.
+                Some(seen_kv) if seen_kv.value != kv.value && (seen_kv.critical || kv.critical) => {
+                    return Err(ParserError::CriticalDuplicateCalendar);
+                }
+                Some(_) => {}
+                None => {
+                    // If the calendar is not yet set, set calendar.
+                    calendar = Some(kv);
+                    continue;
```

Comment:
Nit: With the handler, don't `continue` here any more. Annotations should always reach the handler

Reply:
I do like this approach. I had it initially where every annotation is passed through and talked myself out of it (although perhaps wrong on some level).

What I'm sort of concerned about is the best way to ensure the parser can return a calendar consistently regardless of user code in `handler`. A user could write a handler that undermines the parser on the first calendar instance and only return `None`, in which case the parser won't be able to capture `calendar` even if provided.

The best/most obvious approach would be to just document that calendar's should be returned. A more radical option though may be to remove returning `calendar` as a whole.